Update GitHub Actions: Build Triggers, upload-artifact

k-doering-NOAA opened the pull request:
<!---
Thanks for opening a PR. This commented text will **NOT** appear in the final PR. Toggle between Write and Preview to see what your PR will look like without the comments.
-->

# What is the feature?
@Bai-Li-NOAA and I worked on this to:
* update the build triggers now that we have a dev branch in addition to main.
* Change the upload-artifact github action from v3 to v4. (addresses #718 )

# How have you implemented the solution?
* Made changes to the gha YAML files.

# Does the PR impact any other area of the project, maybe another repo?
* n/a

I was a bit uncertain whether to open this to main or dev, but I think if it gets into dev, and is soon added to main as well, that would be good.

kellijohnson-NOAA commented:
Thanks @k-doering-NOAA and @Bai-Li-NOAA for this PR. I rebased to dev and added one commit. I have a few questions though and will do a review soon.

k-doering-NOAA commented:
> Thanks @k-doering-NOAA and @Bai-Li-NOAA for this PR. I rebased to dev and added one commit. I have a few questions though and will do a review soon.

sounds good!

kellijohnson-NOAA reviewed (approved):
Thank you @k-doering-NOAA and @Bai-Li-NOAA for these changes. I rebased and then made some changes in the following commits (newest to oldest)

* 2555d4ab Refactors GitHub action yml files, puts a comment at the top of the file explaining what the file does. Orders the "on:" to always be pull_request, push, workflow_dispatch. Orders the branch references to be dev then main. Removes some generic comments.
* d743e2d9 fix(call-calc-coverage): Moved and copied the list of file paths to the specific on: calls for pull_request and push because the information underneath each of the following: pull_request, push, and workflow_dispatch, are specific to each trigger. We don't actually want to ignore file paths when someone calls the workflow via dispatch just when things are automatic.
* 5bd6bc11 refactor(build-doxygen): Uses one file instead of two files and an if statement if the workflow should deploy, where deployment happens on main only. @k-doering-NOAA can you review this change to make sure that I didn't mess things up. If you don't like it, I can revert it.
* 6d9d728c fix(clang-format): v18 of workflow

One question I have is if we should be using @v4 or any other calls to versions? I personally think that these always become out of date and we would be better off to just use whatever version is in the main branch. Especially for workflows that are as widely used as actions/upload-artifact@v4.

k-doering-NOAA commented on .github/workflows/call-update-pkgdown.yml:
```diff
@@ -3,9 +3,9 @@
   push:
-    branches: [main, master]
+    branches: [main]
     tags: ['*']
```
@Bai-Li-NOAA and I were also thinking about removing the "tags" part! I think it depends what we want, but if we always want docs to by in sync with main, I think removing the tags part makes sense. Agree that there is nothing wrong with your release protocol!

k-doering-NOAA commented:
Thanks @kellijohnson-NOAA ! I replied to the q re: tags. For https://github.com/NOAA-FIMS/FIMS/commit/5bd6bc11e38eec5b54f2c19dc0a4a02c79d92c38, I think it looks good!

I also think the dependabot action sounds like a good idea (in fact, I should use it for ghactions4r, too...) Let me know if you would like me to remove the tags line or add the dependabot action. I'm also fine if you want to add in those changes!

kellijohnson-NOAA reviewed (approved):
@k-doering-NOAA I removed the one call to tag and added the dependabot file. It is not currently working but my guess is that is because it has yet to be called, I set it to run just once a week rather than daily.
